Fix deprecation warnings

dasm opened the pull request:
Pytest changed the way how type argument should be provided.
Updated to incorporate upstream changes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pytest-dev/pytest-cloud/15)
<!-- Reviewable:end -->

dasm committed:
Fix deprecation warnings

Pytest changed the way how type argument should be provided.
Updated to incorporate upstream changes.

bubenkoff reviewed (changes requested):
the minimum pytest version should be bumped accordingly then

bubenkoff reviewed (changes requested):
and, as usual, please add a changelog entry

dasm commented:
I'm trying to "guesstimate" pytest min version. Maybe you can help me with that.

Changelog for [release 3.0.0](https://docs.pytest.org/en/latest/changelog.html#id1093) (2016-08-18) shows that this change was introduced to `pytest`: https://github.com/pytest-dev/pytest/issues/1740
```
optparse type usage now triggers DeprecationWarnings (#1740).
```
When ran tests with `tox` to verify if it is working for `pytest==2.7.0` I got an error:
```
pkg_resources.VersionConflict: (pytest 2.7.0 (/apps/pytest-cloud/.tox/py27/lib/python2.7/site-packages), Requirement.parse('pytest>=4.4.0'))
ERROR: InvocationError for command /apps/pytest-cloud/.tox/py27/bin/py.test tests --junitxml=/apps/pytest-cloud/.tox/py27/log/junit-py27.xml (exited with code 1)
```
Tox is using
```
tox.ini:7:commands= py.test tests --junitxml={envlogdir}/junit-{envname}.xml
```

Should I assume that minimum version is `pytest>=4.4.0` because it is being used by tox (and automated tests?) Pytest 4.4.0 [release notes](https://docs.pytest.org/en/latest/changelog.html#pytest-4-4-0-2019-03-29) was release 2019-03-29. Or maybe we should assume last supported for python2.7, which would mean `pytest>=4.6.0`?

It might be messed up with my venv too.

Any thougths @bubenkoff

bubenkoff commented:
@dasm in order to test with tox with different pytest, you'd need to remove pytest from requirements-testing.txt and add it in the corresponding env:

```
[testenv]
commands= py.test tests --junitxml={envlogdir}/junit-{envname}.xml
deps =
    -e.
    -r{toxinidir}/requirements-testing.txt
    pytest==2.7.0
passenv = DISPLAY TRAVIS TRAVIS_JOB_ID TRAVIS_BRANCH COVERALLS_REPO_TOKEN USER PWD
```

bubenkoff commented:
@dasm or maybe it will actually work even without touching requirements-testing, as there's no actual conflict

dasm commented:
@bubenkoff I did what you asked for. The error is present, because `pytest-xdist` is depending on `pytest>=4.4.0`: https://github.com/pytest-dev/pytest-xdist/blob/master/setup.py#L3
```
install_requires = ["execnet>=1.1", "pytest>=4.4.0", "pytest-forked", "six"]
```

So, should we just change dependencies list to be dependent on `pytest-xdist` or update both?

dasm commented:
Looks like, this change was introduced with `1.28.0` https://github.com/pytest-dev/pytest-xdist/blob/master/CHANGELOG.rst#pytest-xdist-1280-2019-04-02

dasm commented:
With `pytest==4.4.0` I don't see deprecation warnings and tests are passing.

bubenkoff commented:
@dasm then we should bump the pytest only, as that's the root cause

dasm commented:
I've been thinking about this for some time. Technically, it is not a problem with `pytest>=2.7.0`. It is rather problem with latest `xdist`. `pytest-xdist` has internal dependency on `pytest>=4.4.0`. That's why it is showing an error, when I'm trying to run older version of pytest. Additionally, I found a bug https://github.com/pytest-dev/pytest-cloud/issues/16 with `pytest-xdist>=1.26.0`.

My recommendation would be:

- change `setup.py` for requirements to `pytest-xdist>=1.22.1,<1.26.0`
- remove hard dependency on `pytest>=2.7.0` (or as you suggested, update to the same like in `pytest-xdist`. Disadvantage to tracking `pytest-xdist` is - every time when it is changing, also `pytest-cloud` would need to be updated and released).
- release `pytest-cloud 3.2.0`

After that release, I could start looking into fix for `pytest-xdist>=1.26.0`.

As an alternative, we might include temporary workaround for release, which would be https://github.com/pytest-dev/pytest-cloud/issues/16#issuecomment-618623479

What do you think about that @bubenkoff ?

bubenkoff commented:
so we then have two problems which we can solve with version requirements:

pytest>=2.7.0 -> config typing which you solved already
pytest-xdist>=1.26.0 -> also found by you how to solve

so please implement both in this PR

dasm committed:
Cleanup of supported version

Due to changes in external dependency to pytest-xdist,
pytest-xdist<1.26.0 is not supported anymore.

dasm commented:
@bubenkoff please verify if this change is what you were thinking about.

bubenkoff committed:
Update AUTHORS.rst

bubenkoff reviewed (changes requested):
with the bumping of both pytest and pytest-xdist, it makes sense to raise the major version to 4.0.0

dasm committed:
Addressed review comments

bubenkoff commented on CHANGES.rst:
```diff
@@ -7,7 +7,7 @@ Changelog
 - Fix deprecation warnings for pytest (dasm)
 - Support pytest-xdist >= 1.26.0 (dasm)
-- Support pytest >= 3.6.1 (version supported by pytest-xdist without a bug) (dasm)
+- Support pytest >= 4.0.0 (dasm)
```
sorry, I meant to raise the version of pytest-cloud itself, this change was good

dasm replied:
Oh, makes sense.

dasm committed:
Prepare 4.0.0 release
